### DISCUSSION TITLE
Manually set id of session and use correct user in TokenStorage

### DISCUSSION THREAD

**Discussion by danrot** [symfony/symfony#48194] [Q&A] [answered]
I am currently implementing a payment gateway, and I have some problems with the session, which seems to be caused by a special browser handling.

In my case I want to implement a payment page, which means that the user is directed to a page from the payment gateway, which runs on a different domain. The user enters the credit card information or whatever is used to pay there. After the payment is authorized, the payment provider redirects back to my web application to a defined URL.

So far so good, but now comes the problem: At least Chrome seems not to send the cookies after the redirect from another domain, which causes Symfony to start a new session and in consequence the user is logged out, which is problematic, because I cannot continue the process. From what I read in the documentation the payment provider even suggests not relying on cookies for that matter.

After googling a bit I think the best solution is to introduce a new parameter, store the relation between that parameter and the session id somehow, add the introduced parameter to the query string, and use that information to manually set the session id after the redirect has happened without a cookie.

I have planned to implement that somehow like this:

```php
    #[
        Template(),
        Route('/confirm')
    ]
    public function confirm(Request $request): void
    {
        $sessionId = $this->sessionIdRepository->get($request->query->get('redirectId'));

        $session = $this->requestStack->getSession();
        $session->setId((string) $sessionId);
        // Not sure if that is necessary
        $session->start();

        $this->denyAccessUnlessGranted('IS_AUTHENTICATED_FULLY');
    }
```

But unfortunately that does not seem to work as expected, since the `TokenStorage` still returns an empty token, although the session seems to hold the correct values. I have debugged a bit, and realized that the token is set in the [`ContextListener`](https://github.com/symfony/symfony/blob/6.2/src/Symfony/Component/Security/Http/Firewall/ContextListener.php#L144), which is probably not that easy to trigger again in a `Controller`.

My current guess would be that I have to introduce a listener or subscriber that comes before the `ContextListener` and set the session id there. Actually I wanted to avoid that, because it just affects a single page, and that listener would then be run for every request.

Now my question is: Do I really have to implement a `SetSessionIdListener` or something similar? Or is there another way to handle that solely in the `Controller`, which I am currently missing? And wouldn't you say that the `setId` function of the session should also trigger an update of the `TokenStorage`? Is that something for which you would accept a PR (I don't have the time to that right now, but I would be happy to do that at some point)? It feels to me like I should not be the only one with that problem...

**Comment by MatTheCat** [comment, marked as answer]
Did you configure `framework.session.cookie_samesite`? It should be `lax`.

If Saferpay redirects to your site using a POST request (and you don’t want to [disable SameSite](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie/SameSite#none) for session cookies) then you should create a [custom authenticator](https://symfony.com/doc/current/security/custom_authenticator.html) to authenticate the user against a value of your choice.

**Reply by danrot** [reply]
Now I've already built the workaround and have to finish some work, so I cannot test it 🙈 But the way I see it, this cannot be the solution... As written above, the browser does not send the cookie, not even for the request that goes to the server that set the cookie. 

Additionally payment providers also mention that the session can get lost in the payment process (see e.g. the [Saferpay documentation](https://docs.saferpay.com/home/integration-guide/licences-and-interfaces/payment-page#2-redirect-and-transaction)). So I think using a session should be avoided for this case. That's why I think it would make sense to support switching to another session (and user) in the controller. I am aware that this might open up some security vulnerabilities if implemented wrong, but  the way I see it there is no other possibility to make this work.

**Reply by MatTheCat** [reply]
> Now I've already built the workaround and have to finish some work, so I cannot test it

What do you mean? Don’t you use a versioning system?

> But the way I see it, this cannot be the solution... As written above, the browser does not send the cookie, not even for the request that goes to the server that set the cookie.

> Additionally payment providers also mention that the session can get lost in the payment process (see e.g. the [Saferpay documentation](https://docs.saferpay.com/home/integration-guide/licences-and-interfaces/payment-page#2-redirect-and-transaction)). So I think using a session should be avoided for this case.

I’m baffled you accept and reject solutions without any understanding of a problem. The fact saferpay doesn’t explain anything probably means they don’t know why it is happening either, and I never **ever** heard about such a weird take about session cookies.

Did you read https://blog.chromium.org/2020/02/samesite-cookie-changes-in-february.html? Seems to match your issue pretty well. Can you investigate it before trying to hack your way?

(Anyways I edited my answer to address the case where `Lax` cookies wouldn’t work.)

**Reply by danrot** [reply]
Okay, in the mean time I had time to test it, and lax cookies really seem to work... I've read about the samesite attribute once but never had any issues with it yet, so that didn't come to my mind.

And another thing: It might not have been your intention, but your answer sounded quite condescending to me, which I think is unfair. We are working with version control systems, but nevertheless I had to finish other tasks first. Also, if I don't get any response here within a few days (not that I am entitled to it, I can just post here and hope for the best) I have to implement something to make it work. You also don't know in which state or phase the project currently is, therefore I think you are not capable of evaluating if I have to fix that right away or if it makes sense to finish other tasks first.

I just wanted to give a quick feedback, that I am not able to test it right away, so that you know you don't have to receive an answer right away, and while at it, I added a few more lines that might help others or you to better understand the problem (although that seemed to be wrong).

And I am still not sure it would be a bad idea to solve it without a lax cookie, because introducing that probably exposes the application to other attacks, so working around that with the authenticator as you suggested (which I didn't come up with) might also be a valid way to go.

And I am not just accepting solutions without thinking about it, otherwise I would not have created this discussion in the first place.

**Reply by MatTheCat** [reply]
Nice, wasn’t sure it was related because `lax` is the recipe’s value (that’s why I think it’s safe enough) and I have no idea what the behavior is when the default `null` is used.

Sorry about my tone I did not want to sound condescending. Your problem was a [X/Y one](https://xyproblem.info/) that’s why I pointed your accepted solution (fortunately you did provide enough context to solve it). Also I only asked for you to **check** your `framework.session.cookie_samesite` value so I did not understand why it would take some time :sweat_smile:
